bitswap.go

```diff
@@ -357,7 +344,8 @@ func (bs *Bitswap) ReceiveError(err error) {
 
 // Close is called to shutdown Bitswap
 func (bs *Bitswap) Close() error {
-	return bs.process.Close()
+	bs.cancel()
```

Comment:
Unfortunately, this won't _wait_ for bitswap to stop, that's really why we use goprocess. However, we can do the "go" thing and use a waitgroup.

That is, every where we would have called `process.Go(...)`, call:

```
bs.wg.Add(1)
go func() {
defer bs.wg.Done()
}()
```

(and then `bs.wg.Wait()` in `Close()`).

Reply:
make sense (though I wonder what else we should be waiting to close -- sessions, wantmanager, etc)

will add.

Reply:
Ideally, we'd wait for all workers.

Reply:
So, you know, if we need both a context and a waitgroup, and we need to cover all our workers -- if waiting for shutdowns across the board is really something we should be doing -- I wonder if we should just keep goprocess? I notice you have some recent commits on the repo -- may be we just start maintaining it? Also are we using it elsewhere? Maybe libp2p?

Reply:
Cause otherwise I'm gonna endup writing some code that looks like a mini-goprocess anyway.

Reply:
> I wonder if we should just keep goprocess?

I'd _like_ to (a) switch to something simpler and (b) not leak it across abstraction boundaries.

And yes, we're also using it in go-libp2p.

(also, https://github.com/ipfs/go-ipfs/issues/5810#issuecomment-491462403)

Reply:
@Stebalien ok I built a thing -- living in this repo for now

The main goals were:
- keep it to a single file
- make it flexible to work easily across API boundaries with minimal changes

If this makes sense, after this PR, the next step is to try applying it to the remaining long running processes in go-bitswap

Reply:
see `monitor/monitor.go`